### PR TITLE
#963 specific powershell module version install

### PR DESCRIPTION
**Pull request by prathameshzarkar9**
[Issue 963](https://github.com/devcontainers/features/issues/963)
**Feature Name:**
- Powershell

**Description:**
This PR does the following -
- modules specified in the list has previledges to install specific version of the module which can be mentioned by adding a separator `:` 

_Changelog:_
- Updated install.sh
    - implemented a logic to bifurcate the module and version using a separator `:` in the modules and install the modules latest version if the version is not specified else use the specific version.
- Updated scenarios.json
    - added a version for one of the modules to test the functionality
- Updated devcontainer-feature.json
    - incremented minor version of the feature

**Checklist:**

- [x] It should install the powershell module for the specified version if the version is passed by separator `:`

### CODE REVIEW

File: src/powershell/devcontainer-feature.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
     "id": "powershell",
-    "version": "1.3.5",
+    "version": "1.3.6",
```
Comment (samruddhikhandale) [inline src/powershell/devcontainer-feature.json:3]:
```suggestion
    "version": "1.4.0",
```

File: src/powershell/install.sh [hunk 2]
```diff
@@ -242,14 +242,22 @@ if [ "${use_github}" = "true" ]; then
...
-# If PowerShell modules are requested, loop through and install 
+# If PowerShell modules are requested, loop through and install
 if [ ${#POWERSHELL_MODULES[@]} -gt 0 ]; then
     echo "Installing PowerShell Modules: ${POWERSHELL_MODULES}"
     modules=(`echo ${POWERSHELL_MODULES} | tr ',' ' '`)
     for i in "${modules[@]}"
     do
-        echo "Installing ${i}"
-        pwsh -Command "Install-Module -Name ${i} -AllowClobber -Force -Scope AllUsers" || continue
+        IFS=':' read -ra module_parts <<< "$i"
```
Comment (samruddhikhandale) [inline src/powershell/install.sh:251]: Can we optimize the code? (untested)

```
IFS=':' read -ra module_parts <<< "$i"
module_name="${module_parts[0]}"
args="-Name ${module_name} -AllowClobber -Force -Scope AllUsers"

if [ "${#module_parts[@]}" -eq 2 ]; then
    module_version="${module_parts[1]}"
    echo "Installing ${module_name} version ${module_version}"
    args+=" -RequiredVersion ${module_version}"
else
    echo "Installing module : ${i} , version : latest"
fi

pwsh -Command "Install-Module $args" || continue
```

File: src/powershell/devcontainer-feature.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
...
     "name": "PowerShell",
     "documentationURL": "https://github.com/devcontainers/features/tree/main/src/powershell",
     "description": "Installs PowerShell along with needed dependencies. Useful for base Dockerfiles that often are missing required install dependencies like gpg.",
```
Comment (samruddhikhandale) [inline src/powershell/devcontainer-feature.json:6]: Can we update `modules` description to broadcast this change?

File: src/powershell/install.sh [hunk 4]
```diff
@@ -242,14 +242,22 @@ if [ "${use_github}" = "true" ]; then
...
+        if [ "${#module_parts[@]}" -eq 2 ]; then
+            module_name="${module_parts[0]}"
+            module_version="${module_parts[1]}"
+            echo "Installing ${module_name} version ${module_version}"
```
Comment (samruddhikhandale) [inline src/powershell/install.sh:255]:
```suggestion
            echo "Installing ${module_name}  v${module_version}"
```

File: src/powershell/install.sh [hunk 5]
```diff
@@ -242,14 +242,22 @@ if [ "${use_github}" = "true" ]; then
...
+            pwsh -Command "Install-Module -Name ${module_name} -RequiredVersion ${module_version} -AllowClobber -Force -Scope AllUsers" || continue
+        else
+            echo "Installing module : ${i} , version : latest"
```
Comment (samruddhikhandale) [inline src/powershell/install.sh:258]:
```suggestion
            echo "Installing latest version for ${i} module."
```

File: test/powershell/scenarios.json [hunk 6]
```diff
@@ -3,7 +3,7 @@
     "image": "mcr.microsoft.com/devcontainers/base:jammy",
     "features": {
       "powershell": {
-        "modules": "az.resources, az.storage",
+        "modules": "az.resources:7.1.0, az.storage",
```
Comment (samruddhikhandale) [inline test/powershell/scenarios.json:6]: Can we create a new scenario for testing this change?

File: src/powershell/install.sh (hunk 2 above)
Comment (samruddhikhandale) [inline src/powershell/install.sh:251]:
```suggestion
        IFS='==' read -ra module_parts <<< "$i"
```

Rethinking about this, let's use `==` as the delimiter. We can sync with Python's way of pinning to a version.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/powershell/devcontainer-feature.json (hunk 3 above)
Comment (prathameshzarkar9) [reply src/powershell/devcontainer-feature.json:6]: Done

File: src/powershell/install.sh (hunk 2 above)
Comment (prathameshzarkar9) [reply src/powershell/install.sh:251]: Done

File: test/powershell/scenarios.json (hunk 6 above)
Comment (prathameshzarkar9) [reply test/powershell/scenarios.json:6]: created new test

File: src/powershell/devcontainer-feature.json [hunk 7]
```diff
@@ -18,7 +18,7 @@
         "modules": {
             "type": "string",
             "default": "",
-            "description": "Optional comma separated list of PowerShell modules to install."
+            "description": "Optional comma separated list of PowerShell modules to install. Use delimiter `==` and the version to install specific version."
```
Comment (samruddhikhandale) [inline src/powershell/devcontainer-feature.json:21]:
```suggestion
            "description": "Optional comma separated list of PowerShell modules to install. If you need to install a specific version of a module, use '==' to specify the version (e.g. 'az.resources==2.5.0')"
```

File: test/powershell/scenarios.json [hunk 8]
```diff
@@ -16,5 +16,14 @@
         "powershellProfileURL": "https://raw.githubusercontent.com/codspace/powershell-profile/main/Test-Profile.ps1"
       }
     }
+  },
+  "install_modules_version": {
+    "image": "mcr.microsoft.com/devcontainers/base:jammy",
+    "features": {
+      "powershell": {
+        "modules": "az.resources==2.5.0, az.storage",
+        "powershellProfileURL": "https://raw.githubusercontent.com/codspace/powershell-profile/main/Test-Profile.ps1"
```
Comment (samruddhikhandale) [inline test/powershell/scenarios.json:25]:
```suggestion
```

File: test/powershell/scenarios.json [hunk 9]
```diff
@@ -16,5 +16,14 @@
         "powershellProfileURL": "https://raw.githubusercontent.com/codspace/powershell-profile/main/Test-Profile.ps1"
       }
     }
+  },
+  "install_modules_version": {
+    "image": "mcr.microsoft.com/devcontainers/base:jammy",
+    "features": {
+      "powershell": {
+        "modules": "az.resources==2.5.0, az.storage",
```
Comment (samruddhikhandale) [inline test/powershell/scenarios.json:24]:
```suggestion
        "modules": "az.resources==2.5.0, az.storage"
```

File: test/powershell/install_modules_version.sh [hunk 10]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash
+
+set -e
+
+# Import test library for `check` command
+source dev-container-features-test-lib
+
+# Extension-specific tests
+check "az.resources" pwsh -Command "(Get-Module -ListAvailable -Name Az.Resources).Version.ToString()"
```
Comment (samruddhikhandale) [inline test/powershell/install_modules_version.sh:9]: Can you write a test to validate that the version installed for `az.resources` is `2.5.0`?

File: test/powershell/scenarios.json (hunk 9 above)
Comment (samruddhikhandale) [inline test/powershell/scenarios.json:24]:
```suggestion
        "modules": "az.resources==2.5.0, az.storage==4.3.0",
```

File: test/powershell/install_modules_version.sh [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+check "az.storage" pwsh -Command "(Get-Module -ListAvailable -Name Az.Storage).Version.ToString()"
```
Comment (samruddhikhandale) [inline test/powershell/install_modules_version.sh:10]: Same here

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/powershell/devcontainer-feature.json (hunk 7 above)
Comment (prathameshzarkar9) [reply src/powershell/devcontainer-feature.json:21]: Done

File: src/powershell/install.sh [hunk 12]
```diff
@@ -242,14 +242,25 @@ if [ "${use_github}" = "true" ]; then
...
     echo "Installing PowerShell Modules: ${POWERSHELL_MODULES}"
     modules=(`echo ${POWERSHELL_MODULES} | tr ',' ' '`)
     for i in "${modules[@]}"
     do
-        echo "Installing ${i}"
-        pwsh -Command "Install-Module -Name ${i} -AllowClobber -Force -Scope AllUsers" || continue
+        module_parts=(`echo ${i} | tr '==' ' '`)
+        module_name="${module_parts[0]}"  
+        args="-Name ${module_name} -AllowClobber -Force -Scope AllUsers"  
+        echo "${#module_parts[@]}"
```
Comment (samruddhikhandale) [inline src/powershell/install.sh:254]:
```suggestion
```

File: src/powershell/install.sh [hunk 13]
```diff
@@ -242,14 +242,25 @@ if [ "${use_github}" = "true" ]; then
...
+        if [ "${#module_parts[@]}" -eq 2 ]; then
+            module_version="${module_parts[1]}"
+            echo "Installing ${module_name} v${module_version}.."
```
Comment (samruddhikhandale) [inline src/powershell/install.sh:257]:
```suggestion
            echo "Installing ${module_name} v${module_version}"
```

File: src/powershell/install.sh [hunk 14]
```diff
@@ -242,14 +242,25 @@ if [ "${use_github}" = "true" ]; then
...
+            args+=" -RequiredVersion ${module_version}"
+        else
+            echo "Installing latest version for ${i} module.."
```
Comment (samruddhikhandale) [inline src/powershell/install.sh:260]:
```suggestion
            echo "Installing latest version for ${i} module"
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

**Review by samruddhikhandale — APPROVED** [review]
Wonderful, thank you ⚡

**Pull request merged**